### ISSUE TITLE
Some titles do not work

### ISSUE THREAD

**Issue by CHJ85** [warren-bank/crx-Tubi-TV#1]
Hi there.
Great script.
However, some titles do not work, such as:
https://tubitv.com/series/4829/the-13-ghosts-of-scooby-doo
https://tubitv.com/series/300006855/scooby-doo-and-scrappy-doo
https://tubitv.com/series/300006854/scooby-doo-where-are-you
https://tubitv.com/series/300006856/the-flintstones
https://tubitv.com/series/4846/pebbles-and-bamm-bamm
Could you look into this?

**Comment by warren-bank** [comment]
I agree. I've hit that roadblock a few times, as well.
The issue has to do with videos that are DRM protected.

tldr;
I'll take a deeper look into the feasability for adding DRM support to both the in-browser video players and Chromecast senders.

Using your first link as an example (Scooby-Doo):
* you can view the page source: `view-source:https://tubitv.com/series/4829/the-13-ghosts-of-scooby-doo`
* most of this source is an enormous javascript data structure that is being stored to the global variable: `window.__data`
* the first video in this data structure is represented as the JSON object:
  ```javascript
    {
      "id": "559351",
      "type": "v",
      "title": "S01:E02 - Scooba Kadoobra",
      "description": "Mordor the Malevolent lures the gang into his evil forest in hopes of destroying the only ones who could jeopardize his new-found freedom.",
      "video_resources": [{
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002Fd0193aaa-de2e-4a8c-bb2d-cfe8d5453f06\u002Fng146fqme6.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_widevine_psshv0",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=widevine_psshv0&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiY2dtc19yZXF1aXJlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6InYxIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.GpYbYl9dZYDofZhNhgKa3KseoV6Uc10-2nt1t3F62Dw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_v1"
        }
      }, {
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002Fd0193aaa-de2e-4a8c-bb2d-cfe8d5453f06\u002Fng146fqme6.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_playready_psshv0",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=playready_psshv0&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiY2dtc19yZXF1aXJlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6InYxIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.GpYbYl9dZYDofZhNhgKa3KseoV6Uc10-2nt1t3F62Dw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_v1"
        }
      }, {
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002F1edc68cd-5b49-40b4-a178-9ec10f4c810d\u002F7a555k69fr.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_fairplay",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=fairplay&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiY2dtc19yZXF1aXJlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6InYxIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.GpYbYl9dZYDofZhNhgKa3KseoV6Uc10-2nt1t3F62Dw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_v1"
        }
      }, {
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002Fd0193aaa-de2e-4a8c-bb2d-cfe8d5453f06\u002Fux5hh4yao7.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_widevine_psshv0",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=widevine_psshv0&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiZW5hYmxlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6ImRpc2FibGVkIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.Scue1S1IMur4aYJkIKI7wn5I1g1iu5OcEEi-lurDsNw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_disabled"
        }
      }, {
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002Fd0193aaa-de2e-4a8c-bb2d-cfe8d5453f06\u002Fux5hh4yao7.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_playready_psshv0",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=playready_psshv0&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiZW5hYmxlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6ImRpc2FibGVkIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.Scue1S1IMur4aYJkIKI7wn5I1g1iu5OcEEi-lurDsNw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_disabled"
        }
      }, {
        "manifest": {
          "url": "https:\u002F\u002Fcloudfront.tubi.video\u002F1edc68cd-5b49-40b4-a178-9ec10f4c810d\u002F9ky4tj4xdz.m3u8?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb3VudHJ5IjoiVVMiLCJkZXZpY2VfaWQiOiJkMjRkMjliOC1jNDg0LTQwMzYtYmIwZS04MWQwYTE0ZmE5MDIiLCJleHAiOjE2MjgxMTk4MDAsInBsYXRmb3JtIjoiV0VCIiwidXNlcl9pZCI6MH0.EGMT1v0ODHdm7E1a6M37t8C5dn6quEiQF4OKZF6nNfE&manifest=true",
          "duration": 1422
        },
        "type": "hlsv6_fairplay",
        "license_server": {
          "url": "https:\u002F\u002Flicense.adrise.tv\u002Fchallenge?platform=web&type=fairplay&external_id=69856083-1a71-4c44-a37a-68a49dd9a353+d24d29b8-c484-4036-bb0e-81d0a14fa902&drm_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhbmFsb2dfb3V0IjoiZW5hYmxlZCIsImV4cCI6MTYyODExOTgwMCwiaGRjcCI6ImRpc2FibGVkIiwicHJfc2VjdXJpdHlfbGV2ZWwiOjIwMDAsInd2X3NlY3VyaXR5X2xldmVsIjoxfQ.Scue1S1IMur4aYJkIKI7wn5I1g1iu5OcEEi-lurDsNw",
          "auth_header_key": "customData",
          "auth_header_value": "",
          "hdcp_version": "hdcp_disabled"
        }
      }],
      "subtitles": [{
        "lang": "English",
        "url": "http:\u002F\u002Fs.adrise.tv\u002F39e16ddf-f6f9-46ef-8ca5-eaa351f64460.srt"
      }],
      "url": ""
    }
  ```
* it is parsed by the method: [extract_video_data](https://github.com/warren-bank/crx-Tubi-TV/blob/webmonkey-userscript/es5/webmonkey-userscript/Tubi-TV.user.js#L515)
* when a video is __not__ DRM protected, the _url_ attribute is non-empty.. and contains a URL for an HLS stream (ie: ends with _.m3u8_ extension)
* however, when a video __is__ DRM protected, the _url_ attribute is empty.. and instead the _video_resources_ array includes a list of manifests and the corresponding DRM license server

I'm a little on the fence as to whether to add code to process this array.
* the [ExoAirPlayer](https://github.com/warren-bank/Android-ExoPlayer-AirPlay-Receiver) [sender](http://webcast-reloaded.surge.sh/airplay_sender.html) _does_ already support DRM license credentials
* the rest of the [website](http://webcast-reloaded.surge.sh/) that receives video info:
  - currently supports: video url, captions url, referer url
  - does not currently support: DRM license credentials
* I would need to check to see whether the video players used on the website (ex: Clappr) support:
  - DRM playback in the browser window
  - DRM playback in the corresponding Chromecast receiver app
* if so, then I could:
  - update the Chromecast senders to include this functionality
  - update this userscript for TubiTV to process videos that are only available with DRM encryption:
    * cherry-pick one of the manifests
    * send its URL along with its DRM license credentials to the external website
      - with the option to choose either the _ExoAirPlayer_ sender or the _Chromecast_ sender

**Comment by CHJ85** [comment]
Yea, I thought it was DRM.
Why are you on the fence as to whether to add the code?
If you're able to send its URL along with its DRM license credentials to the external website, I'd say do it − as it would be extremely useful.

**Comment by warren-bank** [comment]
as mentioned, I need to run a few tests to make sure its supported by the players..

if I can manually configure the players with some real data (such as included in the above JSON).. and it works (both in browser and on Chromecast).. then I'll tackle the issue

**Comment by CHJ85** [comment]
Thank you. 
I very much appreciate it.

**Comment by warren-bank** [comment]
quick progress update..

* back on Aug 1, when we originally discussed this issue, I played around with various browser-based video players that support DRM and Chromecast
  - I wasn't able to get any of the drm-protected videos (from your sample links) to play in either:
    * Chrome web browser
    * Chromecast (v1)
  - I pretty much called it a lost cause and gave up

* today, after joking around with a family member who considers "Die Hard" the best movie ever made.. that it was recently added to [Tubi](https://tubitv.com/movies/610703/die-hard?start=true), but then realizing that it was DRM protected..
  - I gave this issue another look
  - this time, focusing only on pushing links to _ExoAirPlayer_.. since widevine (with hdcp disabled) plays nicely on my (very old) test device
  - this userscript now displays all drm links
    * the icon to open a browser tab to the _ExoAirPlayer_ sender webpage embeds all of the necessary DRM info
      - later today, I'll update the target webpage to extract this DRM info and prepopulate the corresponding form fields
    * when running the userscript in _WebMonkey_ on Android
      - the button to play the video passes the DRM info in Intent extras that _ExoAirPlayer_ inspects
      - _ExoAirPlayer_ plays the DRM-protected videos nicely
        * I'm currently watching: _The 13 Ghosts of Scooby-Doo_ S01:E01

**Comment by CHJ85** [comment]
Great work. 
Glad to see some progress here.
However. I have to say, Die Hard is not the best movie ever made. That would have to be Apocalypse Now. 😀
Anyway, thanks for the update.

**Comment by warren-bank** [comment]
OK.. the external website has been updated on both of the domains: https [github pages](https://warren-bank.github.io/crx-webcast-reloaded/external_website/index.html) and http [surge.sh](http://webcast-reloaded.surge.sh/index.html)

however, as mentioned earlier.. support is (currently) limited to playing DRM-protected videos on an Android device using [_ExoAirPlayer_](https://github.com/warren-bank/Android-ExoPlayer-AirPlay-Receiver/releases)

**Comment by CHJ85** [comment]
Thank you!
